features/0454-present-proof-v2/README.md

```diff
@@ -0,0 +1,345 @@
+
+* A similar situation exists with the final transitions for the Prover and its final transitions (Send Presentation Reject, Receive Presentation Ack, and Receive Presentation Reject).
+* When a Prover makes a (counter-)proposal, it transitions to the `proposal-sent` state. This state is only present by implication in the choreography diagram; it essentially equates to the null or begin state in that the Prover does nothing until a presentation request arrives, triggering the leftmost transition for the Prover.
+
+Errors might occur in various places. For example, a Verifier might time out waiting for the Prover to supply a presentation. Errors trigger a `problem-report`. In this version of the protocol, all errors cause the state of both parties (the sender and the receiver of the `problem-report`) to revert to `null` (meaning it is no longer engaged in the protocol at all). Future versions of the protocol may allow more granular choices.
+
+> To Do: Should the previous sentence be addressed in this version? Should it be removed?
+
+### Choreography Diagram
+
+![present proof](credential-presentation.png)
+
+> To Do: Should the Ack or Problem Report be **required** to transition the state to "done"? This is not done in v1.0, and not done in the connection-less presentation use of the v1.0 protocol.
+
+## Messages
+
+The present proof protocol consists of these messages:
+
+* `propose-presentation` - Prover to Verifier (optional) - propose a presentation or send a counter-proposal in response to a `request-presentation` message
+* `request-presentation` - Verifier to Prover - request a presentation
+* `presentation` - Prover to Verifier - provide a presentation in response to a request
+
+In addition, the [`ack`](../0015-acks/README.md) and [`problem-report`](../0035-report-problem/README.md) messages are adopted into the protocol for confirmation and error handling.
+
+The messages that include `~attach` attachments may use either form of the embedded attachment&mdash;`base64` or `json`. In the examples below, one or the other is arbitrarily used.
+
+The `~attach` array is to be used to enable a single presentation to be requested/delivered in different verifiable presentation formats. The ability to have multiple attachments must not be used to request/deliver multiple different presentations in a single instance of the protocol.
+
+### Propose Presentation
+
+An optional message sent by the prover to the verifier to initiate a proof presentation process, or in response to a `request-presentation` message when the prover wants to propose using a different presentation format or request. Schema:
+
+```json
+{
+    "@type": "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/present-proof/%VER/propose-presentation",
+    "@id": "<uuid-propose-presentation>",
+    "comment": "some comment",
+    "formats" : [
+        {
+            "attach_id" : "<attach@id value>",
+            "format" : "<format-and-version>",
```

Comment:
@swcurran the purpose of `format` seems to be to communicate the exact schema of the attachment's contents. This is useful in other protocols as well, so why not add a new field to the attachments RFC?

Reply:
That's an interesting idea that I had not considered.

My mental model for attachments was heavily influenced by email attachments. There, the only typing you need is a MIME type. However, I knew we needed some extra precision in DIDComm, so I added the ability to connect a specific attachment to a specific field in a DIDComm message (e.g., in an insurance claim, you might want to say, "pictures of wind damage on the roof of my shed are in attachment #3" in a machine-readable way). So we have that.

What you're proposing here is attachments getting an additional "format" property that isn't a type, exactly; rather, it's a fine-grained distinction that splits one MIME type into many similar sub-formats. We probably didn't think of this before because we assumed only VC attachments--not photos or office docs or medical data--needed the extra level of detail. If we can satisfy ourselves that it's generally useful, then I guess I'm not opposed to updating the properties of attachments.

I'd be curious to hear Stephen's opinion.

Reply:
@dhh1128 even within VCs we have multiple schemas.

It's not enough to parse a VC generically and validate its proof. Oftentimes we need to parse it and extract data from it, and this is typically from custom, well-defined schemas.

And you're right, attachments can contain anything.

The second point is that this same structure is duplicated in issue-credential V2. So it seems extra-obvious to put this in the attachments RFC.

Reply:
The argument about VC schema doesn't feel convincing to me yet, but maybe say more about it in case I'm missing something? Of course we have different schemas for different VCs -- but we have a mechanism for communicating the schema (@context and/or the credential def from Indy); we don't need to declare it somewhere else. And besides, the value we would put into "format" isn't the schema of the cred anyway; it's something about whether the cred is JSON-LD-centric, JWT-centric, etc -- is it not? 

What would convince me for sure is what I said before: "if we can satisfy ourselves that [the mechanism is] generally useful." And we'd have to carefully explain what we intend by "format" and how it is different from MIME type.

Reply:
@dhh1128 sure - the VC example was poor.

Consider just a JSON object you're trying to get across. MimeType="application/json", format=??

Consider a string representation of something you're sending (eg. a DID (not a did doc)?). MimeType="text/plain", format=??

Reply:
I realized I'm mixing thoughts about how to express the purpose of an attachment vs its format for my use case. I've opened #505 to continue discussion on the purpose of attachments.

I still suspect there's value in this format idea.

> the value we would put into "format" isn't the schema of the cred anyway; it's something about whether the cred is JSON-LD-centric, JWT-centric, etc -- is it not?

It's the schema. From this PR:

> The Hyperledger Indy v1.0 propose filter attachment schema is as follows:

Reply:
To be clear: while I think this format idea probably has merits, this suggestion I've raised is in no way a blocker of this proposal as far as I am concerned.